### ISSUE TITLE
r2d2 is_valid implementation hides the underlying error

### ISSUE THREAD

**Issue by mike-code** [redis-rs/redis-rs#784]
I've been tinkering a bit with how to implement checks on my end if a redis r2d2 connection pool has any alive connections.

I noticed that if I run a redis server with `requirepass` config directive (ie. even "common" commands like `PING` require authentication) then, upon attempt to get a Redis connection from the pool (`my_pool.get()`), the method will return `timed out waiting for connection: broken pipe` error.

That isn't ideal because broken pipe error suggests there's a problem with the TCP connection itself while the real problem is not on the L4 (TCP), but L7 (application) level (missing/invalid password).

Unfortunately the [implementation](https://docs.rs/redis/latest/src/redis/r2d2.rs.html#30) of `r2d2::ManageConnection` trait drops the underlying error and replaces it with standard IO error.

```rust
fn is_valid(&self, conn: &mut Self::Connection) -> Result<(), Self::Error> {
    if conn.check_connection() {
        Ok(())
    } else {
        Err(RedisError::from(io::Error::from(io::ErrorKind::BrokenPipe)))
    }
}
```

My issue with this approach is that if it wasn't for a fact that I tested this scenario on my local machine and that I was doing those tests deliberately (ie. should this issue occur in the wild), one would probably spend a lot of time trying to figure out what causes the TCP connection to fail and blame networking in the first place.

What's your take on that? Am I missing something or have used the pool incorrectly?

**Comment by djc** [comment]
Right, the `check_connection()` interface is probably not ideal for this. Perhaps `is_valid()` should do a `PING` "manually" and then check any `Error` that might be returned to check whether it should considered to be transient or permanent. A relatively simple improvement might be to just check for `ErrorKind::Io`, for example.

Would you be interested in sending a PR along those lines?
